irohad/torii/processor/impl/consensus_status_processor_impl.cpp

```diff
@@ -0,0 +1,122 @@
+ */
+
+#include "torii/processor/consensus_status_processor_impl.hpp"
+
+#include <boost/format.hpp>
+
+#include "interfaces/iroha_internal/block.hpp"
+#include "interfaces/iroha_internal/proposal.hpp"
+#include "interfaces/iroha_internal/transaction_sequence.hpp"
+#include "validation/stateful_validator_common.hpp"
+
+namespace iroha {
+  namespace torii {
+
+    using network::PeerCommunicationService;
+
+    namespace {
+      std::string composeErrorMessage(
+          const validation::TransactionError &tx_error) {
+        if (not tx_error.first.tx_passed_initial_validation) {
+          return (boost::format("Stateful validation error: transaction %s "
+                                "did not pass initial verification: "
+                                "checking '%s', error message '%s'")
+                  % tx_error.second.hex() % tx_error.first.name
+                  % tx_error.first.error)
+              .str();
+        }
+        return (boost::format("Stateful validation error in transaction %s: "
+                              "command '%s' with index '%d' did not pass "
+                              "verification with error '%s'")
+                % tx_error.second.hex() % tx_error.first.name
+                % tx_error.first.index % tx_error.first.error)
+            .str();
+      }
+    }  // namespace
+
+    ConsensusStatusProcessorImpl::ConsensusStatusProcessorImpl(
+        std::shared_ptr<PeerCommunicationService> pcs,
+        std::shared_ptr<iroha::torii::StatusBus> status_bus,
+        std::shared_ptr<shared_model::interface::TxStatusFactory>
+            status_factory)
+        : pcs_(std::move(pcs)),
+          status_bus_(std::move(status_bus)),
+          status_factory_(std::move(status_factory)),
+          log_(logger::log("TxProcessor")) {
+      // process stateful validation results
+      addSubscription(pcs_->on_verified_proposal().subscribe(
+          [this](std::shared_ptr<validation::VerifiedProposalAndErrors>
+                     proposal_and_errors) {
+            handleOnVerifiedProposal(proposal_and_errors);
+          }));
+
+      // commit transactions
+      addSubscription(pcs_->on_commit().subscribe(
+          [this](synchronizer::SynchronizationEvent sync_event) {
+            handleOnCommit(sync_event);
+          }));
+    }
+
+    void ConsensusStatusProcessorImpl::publishStatus(
+        shared_model::interface::TxStatusFactory::TxStatusFactoryInvoker
+            invoker,
+        const shared_model::crypto::Hash &hash,
+        const std::string &error) const {
+      status_bus_->publish((status_factory_.get()->*invoker)(hash, error));
+    }
+
+    void ConsensusStatusProcessorImpl::handleOnVerifiedProposal(
+        std::shared_ptr<iroha::validation::VerifiedProposalAndErrors>
+            proposal_and_errors) {
+      // notify about failed txes
+      const auto &errors = proposal_and_errors->second;
+      std::lock_guard<std::mutex> lock(notifier_mutex_);
```

Comment:
Please state the assumptions about `publishStatus` method, maybe this lock is not required. Current implementation of status bus is thread safe, but the interface of status bus does not tell anything about thread safety either.

Reply:
@kamilsa by itself it is not so good solution, because we don't know the implementation of status bus. So, if we remove lock it means that we are assuming concrete implementation here.